### ISSUE TITLE
Migrate uses of GURL to url::Origin when possible

### ISSUE THREAD

**Issue by mariospr** [brave/brave-browser#19849]
## Description 

As explained in the Chromium documentation [1], _"URLs are often not sufficient for security decisions, since the origin may not be present in the URL (e.g., about:blank), may be tricky to parse (e.g., blob: or filesystem: URLs), or may be opaque despite a normal-looking URL (e.g., the security context may be sandboxed)"_. 

Unfortunately, Brave uses at the moment `GURL` in several places where `url::Origin` should be used instead (e.g. in [`brave::BraveRequestInfo::tab_origin`](https://github.com/brave/brave-core/blob/a87d3e4fcea001df5d24b9ce0ad5087e911cced3/browser/net/url_context.h#L68), and [other places](https://github.com/search?q=%22GURL+origin%22+repo%3Abrave%2Fbrave-core&type=Code&ref=advsearch&l=&l=)), which is becoming an increasingly bigger problem as upstream Chromium is progressively migrating away from GURL as well.

Also, note that this not only affects C++ files but also mojo definitions as well, see [for instance](https://github.com/brave/brave-core/blob/a87d3e4fcea001df5d24b9ce0ad5087e911cced3/components/brave_wallet/common/brave_wallet.mojom#L320):
```
module brave_wallet.mojom;

[...]

interface EthJsonRpcController {
  AddEthereumChain(EthereumChain chain, url.mojom.Url origin) => (string chain_id, bool accepted);
  [...]
}
```
 
Thus, filing this issue to track down the effort of converting as much as possible to `url::Origin` instead of using `GURL` (and maybe cases using strings as well, such as [this one](https://github.com/brave/brave-core/blob/a87d3e4fcea001df5d24b9ce0ad5087e911cced3/components/brave_wallet/common/brave_wallet.mojom#L429)).

In summary, as mentioned in [1], we should use the following datatypes to represent origins:
  * **C++**: `url::Origin` or `blink::SecurityOrigin` (instead of `GURL` or `blink::KURL`).
  * **Mojo**: `url.mojom.Origin` (instead of `url.mojom.Url`).
  * **Java**: `org.chromium.url.Origin` (see also `url::Origin::FromJavaObject` and `url::Origin::CreateJavaObject`).

[1] https://chromium.googlesource.com/chromium/src/+/main/docs/security/origin-vs-url.md

**Comment by mariospr** [comment]
cc @pes10k @diracdeltas @iefremov @goodov

**Comment by pes10k** [comment]
cc @pilgrim-brave
